legacy/src/dnn_backend/dnn_backend_miopen.cpp.inc

```diff
@@ -0,0 +1,1236 @@
+    ConvolutionDescriptor_t const& /*conv_desc*/,
+    TensorDescriptor_t const& /*out_desc*/,
+    ConvFwdAlgo_t const& /*algo*/)
+{
+    return 1 << 30;
+}
+
+size_t GPUDNNBackend::get_conv_bwd_data_workspace_size(
+    Handle_t const& /*handle*/,
+    FilterDescriptor_t const& /*filter_desc*/,
+    TensorDescriptor_t const& /*dy_desc*/,
+    ConvolutionDescriptor_t const& /*conv_desc*/,
+    TensorDescriptor_t const& /*dx_desc*/,
+    ConvBwdDataAlgo_t const& /*algo*/)
+{
+    return 1 << 30;
+}
+
+size_t GPUDNNBackend::get_conv_bwd_filter_workspace_size(
+    Handle_t const& /*handle*/,
+    TensorDescriptor_t const& /*in_desc*/,
+    TensorDescriptor_t const& /*dy_Desc*/,
+    ConvolutionDescriptor_t const& /*conv_Desc*/,
+    FilterDescriptor_t const& /*dw_desc*/,
+    ConvBwdFilterAlgo_t const& /*algo*/)
+{
+    return 1 << 30;
+}
+
+void GPUDNNBackend::apply_fwd_bias(Handle_t handle,
+                                   void const* alpha,
+                                   TensorDescriptor_t const& bias_desc,
+                                   void const* const bias,
+                                   void const* beta,
+                                   TensorDescriptor_t const& y_desc,
+                                   void* const y)
+{
+    // FIXME: Does this also need to handle alpha!=1, beta!=0??
+    DISTCONV_CHECK_MIOPEN(miopenConvolutionForwardBias(
+        handle, alpha, bias_desc, bias, beta, y_desc, y));
+}
+
+void GPUDNNBackend::apply_bwd_bias(Handle_t handle,
+                                   void const* alpha,
+                                   TensorDescriptor_t const& dy_desc,
+                                   void const* dy_data,
+                                   void const* beta,
+                                   TensorDescriptor_t const& db_desc,
+                                   void* const db_data)
+{
+    // FIXME: Does this also need to handle alpha!=1, beta!=0??
+    DISTCONV_CHECK_MIOPEN(miopenConvolutionBackwardBias(
+        handle, alpha, dy_desc, dy_data, beta, db_desc, db_data));
+}
+
+auto GPUDNNBackend::get_fwd_algorithm_by_name(std::string const& name)
+    -> ConvFwdAlgo_t
+{
+    return MIOpenConvolutionFwdAlgorithms::get_algo(name);
+}
+
+// MIOpen supports "Find" mode and "Immediate" mode. The "Find" mode
+// REQUIRES that miopenFindConvolution* be called FOR EACH (distinct)
+// convolution that will be run. "Immediate" mode uses "solutions" via
+// an API that isn't easily mapped to this interface, but is closer to
+// the heuristic approach that, e.g., cuDNN supports. So we just error
+// out here.
+auto GPUDNNBackend::get_fwd_algorithm_by_heuristics(
+    Handle_t handle,
+    TensorDescriptor_t input_desc,
+    TensorDescriptor_t filter_desc,
+    ConvolutionDescriptor_t conv_desc,
+    TensorDescriptor_t output_desc,
+    size_t ws_size) -> ConvFwdAlgo_t
+{
+    throw std::runtime_error(
+        "Distconv-MIOpen does not support heuristic algorithm selection.");
```

Comment:
```suggestion
        "DiHydrogen-MIOpen does not support heuristic algorithm selection.");
```

Reply:
That was intentional, but I'm not strongly tied to it. Colloquially we've been referring to this legacy code as "Distconv" but the forthcoming larger refactor/rewrite as "DiHydrogen". There's a path for more heuristic algorithm selection that would be compatible with the latter effort, but would be difficult to shim into such a cuDNN-centric code as DistConv is.